PSReadLine/StringBuilderLinewiseExtensions.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Text;
+
+namespace Microsoft.PowerShell
+{
+    public class Range
+    {
+        public int Offset { get; set; }
+        public int Count { get; set; }
+    }
+
+    public static class StringBuilderLinewiseExtensions
+    {
+        /// <summary>
+        /// Determines the offset and the length of the fragment
+        /// in the specified buffer that corresponds to a
+        /// given number of lines starting from the specified line index
+        /// </summary>
+        /// <param name="buffer" />
+        /// <param name="lineIndex" />
+        /// <param name="lineCount" />
+        public static Range GetRange(this StringBuilder buffer, int lineIndex, int lineCount)
+        { 
+            var length = buffer.Length;
+
+            var startPosition = 0;
+            var startPositionIdentified = false;
+
+            var endPosition = length - 1;
+            var endPositionIdentified = false;
+
+            var currentLine = 0;
+
+            for (var position = 0; position < length; position++)
+            {
+                if (currentLine == lineIndex && !startPositionIdentified)
+                {
+                    startPosition = position;
+                    startPositionIdentified = true;
+                }
+
+                if (buffer[position] == '\n')
+                {
+                    currentLine++;
+                }
+
+                if (currentLine == lineIndex + lineCount && !endPositionIdentified)
+                {
+                    endPosition = position;
+                    endPositionIdentified = true;
+                }
+            }
+
+            return new Range

```

Comment:
This method supports a scenario whereby the user specifies a number of lines to be taken in consideration for yanking.

The `lineCount` argument must not be 0, but is otherwise unbounded. If the user specifies more lines than there are available, making the requested range of lines larger than what it can be, I still want this method to return a valid range of offsets into the buffer.
For instance, if you go to the next to last line of a file in vim and ask <kbd>42</kbd>, <kbd>yy</kbd>, this would nevertheless successfully yank two lines of text in the register.

So I could add a check that `lineCount` is greater than 0.

The `lineIndex` argument comes from the cursor position via a call to the `GetLogicalLineNumber()` so is guaranteed to be valid, and thus, this method safe to return a valid range of offsets into the buffer.

Are you OK for an extra check on `lineCount` at the start of the method, an `Debug.Assert(true == startPositionIdentified)` just before returning the range and more comments in the method documentation?

Reply:
As a general purpose method (extension to StringBuilder), it seems wrong.

As a special purpose method with good comments explaining such boundary conditions, it's fine.

I would make sure `lineCount == 0` is handled reasonably - e.g. user types <kbd>0yy</kbd> or something like that. We don't want exceptions in release builds, so assertions might not be enough.

Reply:
Sure. Thanks for that.